derp/derp_test.go

```diff
@@ -851,125 +851,189 @@ func TestClientSendPong(t *testing.T) {
+		if !ok {
+			t.Error("no set for key")
+			return
 		}
-
-		cin, err := ln.Accept()
-		if err != nil {
-			t.Fatal(err)
+		got := set.ActiveClient()
+		if got != want {
+			t.Errorf("active client = %q; want %q", clientName[got], clientName[want])
 		}
-
-		brwServer := bufio.NewReadWriter(bufio.NewReader(cin), bufio.NewWriter(cin))
-		go s.Accept(cin, brwServer, fmt.Sprintf("test-client-%d", connNum))
-
-		brw := bufio.NewReadWriter(bufio.NewReader(cout), bufio.NewWriter(cout))
-		c, err := NewClient(priv, cout, brw, logger)
-		if err != nil {
-			t.Fatalf("client %d: %v", connNum, err)
+	}
+	checkDup := func(c *sclient, want bool) {
+		t.Helper()
+		if got := c.isDup.Get(); got != want {
+			t.Errorf("client %q isDup = %v; want %v", clientName[c], got, want)
 		}
-		return c
 	}
-
-	wantVar := func(v *expvar.Int, want int64) {
+	checkDisabled := func(c *sclient, want bool) {
 		t.Helper()
-		if got := v.Value(); got != want {
-			t.Errorf("got %d; want %d", got, want)
+		if got := c.isDisabled.Get(); got != want {
+			t.Errorf("client %q isDisabled = %v; want %v", clientName[c], got, want)
 		}
 	}
-
-	wantClosed := func(c *Client) {
+	wantDupConns := func(want int) {
 		t.Helper()
-		for {
-			m, err := c.Recv()
-			if err != nil {
-				t.Logf("got expected error: %v", err)
-				return
-			}
-			switch m.(type) {
-			case ServerInfoMessage:
-				continue
-			default:
-				t.Fatalf("client got %T; wanted an error", m)
-			}
+		if got := s.dupClientConns.Value(); got != int64(want) {
+			t.Errorf("dupClientConns = %v; want %v", got, want)
 		}
 	}
-
-	c1 := connect()
-	waitConnect(t, c1)
-	c2 := connect()
-	waitConnect(t, c2)
-	wantVar(&s.clientsReplaced, 1)
-	wantClosed(c1)
-
-	for i := 0; i < 100+5; i++ {
-		c := connect()
-		defer c.nc.Close()
-		if s.clientsReplaceLimited.Value() == 0 && i < 90 {
-			continue
+	wantDupKeys := func(want int) {
+		t.Helper()
+		if got := s.dupClientKeys.Value(); got != int64(want) {
+			t.Errorf("dupClientKeys = %v; want %v", got, want)
 		}
-		t.Logf("for %d: replaced=%d, limited=%d, sleeping=%d", i,
-			s.clientsReplaced.Value(),
-			s.clientsReplaceLimited.Value(),
-			s.clientsReplaceSleeping.Value(),
-		)
 	}
 
-	mu.Lock()
-	defer mu.Unlock()
-	if sleeps == 0 {
-		t.Errorf("no sleeps")
-	}
-	if slept == 0 {
-		t.Errorf("total sleep duration was 0")
-	}
+	// Common case: a single client comes and goes, with no dups.
+	run("one_comes_and_goes", func(t *testing.T) {
+		wantNoClient()
+		s.registerClient(c1)
+		wantSingleClient(c1)
+		s.unregisterClient(c1)
+		wantNoClient()
+	})
+
+	// A still somewhat common case: a single client was
+	// connected and then their wifi dies or laptop closes
+	// or they switch networks and connect from a
+	// different network. They have two connections but
+	// it's not very bad. Only their new one is
+	// active. The last one, being dead, doesn't send and
+	// thus the new one doesn't get disabled.
+	run("small_overlap_replacement", func(t *testing.T) {
```

Comment:
Hm, this test makes me realize there's still a small race here: a client could send then immediately switch networks and reconnect. Since the sclients aren't synchronized with each other, the derp server could process c2 first, mark it active, then receive a frame from c1 that was still in flight, and incorrectly disable c2 :/

Maybe there needs to be some kind of grace period immediately after c2 shows up, where we still allow a little crosstalk? Distributed systems are annoying :(

Reply:
given how many RTTs a DERP setup takes, I'm not particularly worried about that